File: lista2/q3/main.go
package main

import (
	"time"
	"os"
	"strconv"
)

func main(){
	numberOfThreads, _ := strconv.Atoi(os.Args[1])
	if numberOfThreads == 0 { // to evaluate memory with only one goroutine
		sleep()
	} else {
		gateway(numberOfThreads)
	}
}

func request(c chan int) {
	sleptTime := sleep()
	c <- sleptTime
}

func gateway(num_replicas int) {
    ch0 := make(chan int)
    for i := 0; i < num_replicas; i++ {
	go request(ch0)
    }

    for i := 0; i < num_replicas; i++ {
        <- ch0
    }
}

func sleep() int {
	numberOfSecondsToSleep := 5
	time.Sleep(time.Duration(numberOfSecondsToSleep) * time.Second)
	return numberOfSecondsToSleep
}

File: lista2/q3/Main.java
import java.util.Random;

public class Main {

    public static void main(String args[]) {
        int numberOfThreads = Integer.parseInt(args[0]);
        if (numberOfThreads == 0) {  // to evaluate memory with only one thread
            sleep();
        } else {
            gateway(numberOfThreads);
        }
    }

    public static void gateway(int numReplicas) {
        Request request = new Request();
        Thread[] threads = new Thread[numReplicas];
        for (int i = 0; i < numReplicas; i++) {
            Thread thread = new Thread(request, "request" + i);
            thread.start();
            threads[i] = thread;
        }

        try {
            for (int i = 0; i < numReplicas; i++) {
                threads[i].join();
            }
        } catch (InterruptedException e) {}
    }

    public static class Request implements Runnable {
        public void run() {
            sleep();
        }
    }

    public static void sleep() {
        try {
            int numberOfSecondsToSleep = 5;
            Thread.sleep(numberOfSecondsToSleep * 1000); // Thread.sleep sleeps milliseconds.
        } catch (InterruptedException e) {}
    }
}
